File: external-api/src/main/java/com/engine/external/project/event/EventManagerImpl.java
package com.engine.external.project.event;

import com.engine.external.project.event.events.AbstractEvent;
import com.engine.external.project.event.events.EventHolder;
import reactor.core.Disposable;
import reactor.core.publisher.*;
import reactor.core.scheduler.Scheduler;
import reactor.core.scheduler.Schedulers;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class EventManagerImpl implements EventManager {

    private final Map<EventHolder, Disposable> listeners = new HashMap<>();

    private final FluxProcessor<AbstractEvent, ? super AbstractEvent> abstractEventFluxProcessor = Processors.multicast();
    private final Scheduler scheduler = Schedulers.newSingle("Event Manager");

    private final Sinks.StandaloneFluxSink<Object> eventSink = Sinks.multicast();
    @Override
    public void process(AbstractEvent event) {
        eventSink.next(event);
    }

    @Override
    public <T extends AbstractEvent> Flux<T> on(Class<T> klass) {
        return abstractEventFluxProcessor.publishOn(scheduler)
                .ofType(klass);
    }

    @Override
    public void registerListener(EventHolder holder) {
        on(AbstractEvent.class).subscribe(holder::onEvent);
    }

    @Override
    public Collection<EventHolder> getListeners() {
        return listeners.keySet();
    }

    @Override
    public void unregisterListener(EventHolder holder) {
        listeners.remove(holder).dispose();
    }

}

File: external-sdk/src/main/kotlin/com/engine/external/project/events/EventManagerImpl.kt
package com.engine.external.project.events

import com.engine.external.project.event.EventManager
import com.engine.external.project.event.models.AbstractEvent
import com.engine.external.project.event.models.EventHolder
import reactor.core.Disposable
import reactor.core.publisher.Flux
import reactor.core.publisher.FluxProcessor
import reactor.core.publisher.Processors
import reactor.core.publisher.Sinks
import reactor.core.scheduler.Schedulers

class EventManagerImpl : EventManager {

    private val listeners: MutableMap<EventHolder, Disposable> = HashMap()
    private val scheduler = Schedulers.newSingle("Event Scheduler")
    private val fluxProcessor: FluxProcessor<AbstractEvent, in AbstractEvent> = Processors.multicast()
    private val eventSink = Sinks.multicastNoWarmup<Any>()

    override fun process(event: AbstractEvent) {
        eventSink.next(event)
    }

    override fun <T : AbstractEvent?> on(klass: Class<T>): Flux<T> {
        return fluxProcessor.publishOn(scheduler)
            .log()
            .ofType(klass)
    }

    override fun registerListener(holder: EventHolder) {
        listeners[holder] = on(AbstractEvent::class.java).subscribe { event: AbstractEvent? ->
            holder.onEvent(event)
        }
    }

    override fun getListeners(): Collection<EventHolder> {
        return listeners.keys
    }

    override fun unregisterListener(holder: EventHolder) {
        listeners.remove(holder)!!.dispose()
    }
}
